### PR TITLE
PT-3045: Facets for multivalue, multilanguage properties

### PR DESCRIPTION
**Pull request by akak1977**
## Description
Proper facets handling for multivalue, multilanguage properties.
Add localized label to the facet. Repair facet term label to show proper value accordingly to passed culture.
image

### Jira-link:
https://virtocommerce.atlassian.net/browse/PT-3044

### CODE REVIEW

File: src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Mapping/FacetMappingProfile.cs [hunk 1]
```diff
@@ -13,41 +13,45 @@ public FacetMappingProfile()
...
+                            Label = x.Value.ToString(),
+                        })
                             .ToArray(),
                         Name = request.Field,
                     },
                     _ => null
                 };
-
+                if (result != null)
+                    result.Label = request.Labels?.FirstOrDefault(x => x.Language == cultureName)?.Label ?? result.Name;
```
Comment (Vectorfield4) [inline src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Mapping/FacetMappingProfile.cs:54]:
```suggestion

                if (result != null)
                    result.Label = request.Labels?.FirstOrDefault(x => x.Language == cultureName)?.Label ?? result.Name;
                    
```

File: src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Mapping/FacetMappingProfile.cs [hunk 2]
```diff
@@ -13,41 +13,45 @@ public FacetMappingProfile()
...
-                                Term = x.Value.ToString(),
-                                Label = x.Labels != null ? string.Join(' ', x.Labels.Select(l => l.Label)) : x.Value.ToString(),
-                            })
+                        Terms = request.Items.Select(x => new FacetTerm
+                        {
+                            Count = x.Count,
+                            IsSelected = x.IsApplied,
+                            Term = x.Value.ToString(),
+
+                            Label = x.Labels?.FirstOrDefault(x => x.Language == cultureName)?.Label ?? x.Value.ToString(),
```
Comment (Vectorfield4) [inline src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Mapping/FacetMappingProfile.cs:29]:
```suggestion
                            Label = x.Labels?.FirstOrDefault(x => x.Language == cultureName)?.Label ?? x.Value.ToString(),
```

**Review by Vectorfield4 — APPROVED** [review]

**Review by ksavosteev — APPROVED** [review]

**Review by tatarincev — COMMENTED** [review]
Please do not merge without my review

**Comment by akak1977** [comment]
> Please do not merge without my review

OK

File: src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Queries/SearchProductQueryHandler.cs [hunk 3]
```diff
@@ -103,7 +130,10 @@ public virtual async Task<SearchProductResponse> Handle(SearchProductQuery reque
                 Currency = currency,
                 Store = store,
                 Results = products,
-                Facets = resultAggregations?.Select(x => _mapper.Map<FacetResult>(x)).ToList(),
+                Facets = resultAggregations?.Select(x => _mapper.Map<FacetResult>(x, options =>
```
Comment (tatarincev) [inline src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Queries/SearchProductQueryHandler.cs:133]: The better solution will be to filter the resulting "synthetic" facets for mutilating properties according to the passed culture in this line. All code above can be removed.
e.g
`my-cool-property, my-cool-property_en_us, "my-cool-property_de_de` -> left only single value `my-cool-property` with values from `my-cool-property_en_us`

File: src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Mapping/FacetMappingProfile.cs (hunk 1 above)
Comment (tatarincev) [reply src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Mapping/FacetMappingProfile.cs:54]: All logic for selecting a localized string for a given culture  must be reused from a single place 

https://github.com/VirtoCommerce/vc-module-experience-api/blob/dev/src/VirtoCommerce.ExperienceApiModule.Core/Extensions/IHasLanguageExtensions.cs

You have to extend this extension method to be able to work from object types that not implement `IHasLanguage` interface.
```C#
 public static IHasLanguage FirstBestMatchForLanguage(this IEnumerable hasLanguages, Func<object, string> langSelector, string language)
```

**Review by tatarincev — CHANGES_REQUESTED** [review]

**Pull request merged**